File: scripts/sparse_cholesky/sc_seq.hs
{-# LANGUAGE BangPatterns #-}
module Main where

-- Dense block Cholesky sequential baseline in Haskell — raw Ptr Double,
-- peekElemOff / pokeElemOff in inner kernels. Same algorithm as sc_seq.c.
-- Block (i, j) at element offset blockIdx(i,j) * B*B into the contiguous
-- Ptr Double; element (r, c) of that block at +r*B+c.

import Control.Monad (forM_, when)
import Data.Bits ((.&.))
import Data.Int (Int32)
import Data.Word (Word8, Word32, Word64)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Internal as BSI
import Foreign.ForeignPtr (withForeignPtr)
import Foreign.Marshal.Alloc (mallocBytes)
import Foreign.Marshal.Utils (copyBytes)
import Foreign.Storable (peekElemOff, pokeElemOff)
import Foreign.Ptr (Ptr, castPtr)
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import System.IO (hFlush, stdout, hPutStrLn, stderr)
import System.Environment (getArgs)

readConfig :: FilePath -> IO (Int, Int)
readConfig path = do
  s <- readFile path
  let kv = [ (head ws, read (ws !! 1) :: Int)
           | l <- lines s, let ws = words l, length ws >= 2 ]
  return (maybe 0 id (lookup "NB" kv), maybe 64 id (lookup "B" kv))

{-# INLINE blockIdx #-}
blockIdx :: Int -> Int -> Int
blockIdx i j = i * (i + 1) `div` 2 + j

readBin :: FilePath -> Int -> IO (Ptr Double)
readBin path nbytes = do
  bs <- BS.readFile path
  let !(BSI.BS fp _) = bs
  p <- mallocBytes nbytes :: IO (Ptr Word8)
  withForeignPtr fp $ \src -> copyBytes p src nbytes
  return (castPtr p)

-- POTRF on block at base offset `tgt` (in elements), block size B.
{-# INLINE potrfBlock #-}
potrfBlock :: Ptr Double -> Int -> Int -> IO ()
potrfBlock !arr !tgt !b = do
  forM_ [0..b-1] $ \j -> do
    !s0 <- peekElemOff arr (tgt + j*b + j)
    let go !kk !acc
          | kk >= j   = return acc
          | otherwise = do
              !x <- peekElemOff arr (tgt + j*b + kk)
              go (kk + 1) (acc - x*x)
    !s <- go 0 s0
    let !sq = sqrt s
    pokeElemOff arr (tgt + j*b + j) sq
    let !inv = 1.0 / sq
    forM_ [j+1..b-1] $ \i -> do
      !t0 <- peekElemOff arr (tgt + i*b + j)
      let go2 !kk !acc
            | kk >= j   = return acc
            | otherwise = do
                !a <- peekElemOff arr (tgt + i*b + kk)
                !c <- peekElemOff arr (tgt + j*b + kk)
                go2 (kk + 1) (acc - a*c)
      !t <- go2 0 t0
      pokeElemOff arr (tgt + i*b + j) (t * inv)
  forM_ [0..b-1] $ \i ->
    forM_ [i+1..b-1] $ \j -> pokeElemOff arr (tgt + i*b + j) 0.0

{-# INLINE trsmBlock #-}
trsmBlock :: Ptr Double -> Int -> Int -> Int -> IO ()
trsmBlock !arr !x !l !b =
  forM_ [0..b-1] $ \i ->
    forM_ [0..b-1] $ \j -> do
      !s0 <- peekElemOff arr (x + i*b + j)
      let go !kk !acc
            | kk >= j   = return acc
            | otherwise = do
                !a <- peekElemOff arr (x + i*b + kk)
                !c <- peekElemOff arr (l + j*b + kk)
                go (kk + 1) (acc - a*c)
      !s <- go 0 s0
      !lj <- peekElemOff arr (l + j*b + j)
      pokeElemOff arr (x + i*b + j) (s / lj)

{-# INLINE syrkBlock #-}
syrkBlock :: Ptr Double -> Int -> Int -> Int -> IO ()
syrkBlock !arr !c !a !b =
  forM_ [0..b-1] $ \i ->
    forM_ [0..i] $ \j -> do
      let go !kk !acc
            | kk >= b   = return acc
            | otherwise = do
                !x <- peekElemOff arr (a + i*b + kk)
                !y <- peekElemOff arr (a + j*b + kk)
                go (kk + 1) (acc + x*y)
      !s <- go 0 0.0
      !c0 <- peekElemOff arr (c + i*b + j)
      pokeElemOff arr (c + i*b + j) (c0 - s)

{-# INLINE gemmBlock #-}
gemmBlock :: Ptr Double -> Int -> Int -> Int -> Int -> IO ()
gemmBlock !arr !c !a !b_ !b =
  forM_ [0..b-1] $ \i ->
    forM_ [0..b-1] $ \j -> do
      let go !kk !acc
            | kk >= b   = return acc
            | otherwise = do
                !x <- peekElemOff arr (a + i*b + kk)
                !y <- peekElemOff arr (b_ + j*b + kk)
                go (kk + 1) (acc + x*y)
      !s <- go 0 0.0
      !c0 <- peekElemOff arr (c + i*b + j)
      pokeElemOff arr (c + i*b + j) (c0 - s)

-- (kind, ti, tj, s1i, s1j, s2i, s2j)
readDag :: FilePath -> IO [(Int, Int, Int, Int, Int, Int, Int)]
readDag path = do
  bs <- BS.readFile path
  let !(BSI.BS fp _) = bs
  withForeignPtr fp $ \src8 -> do
    let src = castPtr src8 :: Ptr Int32
    !n <- (fromIntegral :: Int32 -> Int) <$> peekElemOff src 0
    let readOp !off = do
          !k  <- (fromIntegral :: Int32 -> Int) <$> peekElemOff src off
          !ti <- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 1)
          !tj <- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 2)
          !s1i<- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 3)
          !s1j<- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 4)
          !s2i<- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 5)
          !s2j<- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 6)
          !nd <- (fromIntegral :: Int32 -> Int) <$> peekElemOff src (off + 8)
          return ((k, ti, tj, s1i, s1j, s2i, s2j), off + 9 + nd)
    let collect !off !acc !left
          | left == 0 = return (reverse acc)
          | otherwise = do
              (op, off') <- readOp off
              collect off' (op : acc) (left - 1)
    collect 1 [] n

main :: IO ()
main = do
  args <- getArgs
  case args of
    (dir:_) -> run dir
    _       -> hPutStrLn stderr "usage: sc_seq_hs DATA_DIR"

run :: FilePath -> IO ()
run dir = do
  (nb, b) <- readConfig (dir ++ "/config.txt")
  let !nBlocks = nb * (nb + 1) `div` 2
      !blockSize = b * b
      !nbytes = nBlocks * blockSize * 8
      blockOff i j = blockIdx i j * blockSize  -- element offset
  arr <- readBin (dir ++ "/A.bin") nbytes
  ops <- readDag (dir ++ "/dag.bin")

  t0 <- getCurrentTime
  forM_ ops $ \(kind, ti, tj, s1i, s1j, s2i, s2j) ->
    case kind of
      0 -> potrfBlock arr (blockOff ti tj) b
      1 -> trsmBlock  arr (blockOff ti tj) (blockOff s1i s1j) b
      2 -> syrkBlock  arr (blockOff ti tj) (blockOff s1i s1j) b
      3 -> gemmBlock  arr (blockOff ti tj) (blockOff s1i s1j) (blockOff s2i s2j) b
      _ -> error "unknown op"
  t1 <- getCurrentTime
  let !secs = realToFrac (diffUTCTime t1 t0) :: Double

  -- Checksum
  let totalElems = nBlocks * blockSize
      go !i !acc
        | i >= totalElems = return ((acc :: Word64) .&. 0xFFFFFFFF)
        | otherwise = do
            !v <- peekElemOff arr i
            let !fixed = truncate (v * 1e6) :: Int   -- toward zero, matches C (int64_t) cast
                !w = fromIntegral (fromIntegral fixed :: Word32) :: Word64
            go (i + 1) ((acc + w) .&. 0xFFFFFFFF)
  !cs <- go 0 0
  putStrLn ("CHECKSUM=" ++ show cs)
  putStrLn ("RUNTIME_SEC=" ++ show secs)
  hFlush stdout

File: scripts/sparse_cholesky/sc_seq.rs
//! Dense block Cholesky sequential baseline in Rust — raw `*mut f64` pointers,
//! `unsafe` inner loops, matches sc_seq.c algorithm byte-for-byte.

use std::env;
use std::fs::File;
use std::io::Read;
use std::time::Instant;

fn load_config(dir: &str) -> (usize, usize) {
    let path = format!("{}/config.txt", dir);
    let text = std::fs::read_to_string(&path).unwrap();
    let mut nb = 0usize; let mut b = 64usize;
    for line in text.lines() {
        let mut it = line.split_whitespace();
        let key = it.next().unwrap_or("");
        let val: usize = it.next().unwrap_or("0").parse().unwrap_or(0);
        match key { "NB" => nb = val, "B" => b = val, _ => {} }
    }
    (nb, b)
}

#[inline(always)] fn block_idx(i: usize, j: usize) -> usize { i * (i + 1) / 2 + j }

unsafe fn potrf_block(d: *mut f64, b: usize) {
    for j in 0..b {
        let mut s = *d.add(j*b + j);
        for kk in 0..j { let x = *d.add(j*b + kk); s -= x * x; }
        let sqrt_s = s.sqrt();
        *d.add(j*b + j) = sqrt_s;
        let inv = 1.0 / sqrt_s;
        for i in (j+1)..b {
            let mut t = *d.add(i*b + j);
            for kk in 0..j { t -= *d.add(i*b + kk) * *d.add(j*b + kk); }
            *d.add(i*b + j) = t * inv;
        }
    }
    // Zero upper triangle
    for i in 0..b { for j in (i+1)..b { *d.add(i*b + j) = 0.0; } }
}

unsafe fn trsm_block(x: *mut f64, l: *const f64, b: usize) {
    for i in 0..b {
        for j in 0..b {
            let mut s = *x.add(i*b + j);
            for kk in 0..j { s -= *x.add(i*b + kk) * *l.add(j*b + kk); }
            *x.add(i*b + j) = s / *l.add(j*b + j);
        }
    }
}

unsafe fn syrk_block(c: *mut f64, a: *const f64, b: usize) {
    for i in 0..b {
        for j in 0..=i {
            let mut s = 0.0;
            for kk in 0..b { s += *a.add(i*b + kk) * *a.add(j*b + kk); }
            *c.add(i*b + j) -= s;
        }
    }
}

unsafe fn gemm_block(c: *mut f64, a: *const f64, b_: *const f64, b: usize) {
    for i in 0..b {
        for j in 0..b {
            let mut s = 0.0;
            for kk in 0..b { s += *a.add(i*b + kk) * *b_.add(j*b + kk); }
            *c.add(i*b + j) -= s;
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 { eprintln!("usage: {} DATA_DIR", args[0]); std::process::exit(2); }
    let dir = &args[1];
    let (nb, b) = load_config(dir);
    let n_blocks = nb * (nb + 1) / 2;
    let block_size = b * b;

    // Load A.bin
    let path = format!("{}/A.bin", dir);
    let mut f = File::open(&path).unwrap();
    let mut buf = vec![0u8; n_blocks * block_size * 8];
    f.read_exact(&mut buf).unwrap();
    let mut a: Vec<f64> = (0..n_blocks * block_size).map(|i| {
        let off = i * 8;
        f64::from_le_bytes([buf[off], buf[off+1], buf[off+2], buf[off+3],
                            buf[off+4], buf[off+5], buf[off+6], buf[off+7]])
    }).collect();

    // Load DAG
    let dag_path = format!("{}/dag.bin", dir);
    let mut df = File::open(&dag_path).unwrap();
    let mut hdr4 = [0u8; 4];
    df.read_exact(&mut hdr4).unwrap();
    let n_ops = i32::from_le_bytes(hdr4) as usize;

    struct Op { kind: i32, ti: usize, tj: usize, s1i: usize, s1j: usize, s2i: usize, s2j: usize }
    let mut ops: Vec<Op> = Vec::with_capacity(n_ops);
    for _ in 0..n_ops {
        let mut hdr = [0u8; 36];
        df.read_exact(&mut hdr).unwrap();
        let vals: [i32; 9] = std::array::from_fn(|i|
            i32::from_le_bytes([hdr[i*4], hdr[i*4+1], hdr[i*4+2], hdr[i*4+3]]));
        let n_deps = vals[8] as usize;
        if n_deps > 0 {
            let mut skip = vec![0u8; n_deps * 4];
            df.read_exact(&mut skip).unwrap();
        }
        ops.push(Op {
            kind: vals[0], ti: vals[1] as usize, tj: vals[2] as usize,
            s1i: vals[3].max(0) as usize, s1j: vals[4].max(0) as usize,
            s2i: vals[5].max(0) as usize, s2j: vals[6].max(0) as usize,
        });
    }

    let t0 = Instant::now();
    unsafe {
        let a_ptr = a.as_mut_ptr();
        for op in &ops {
            let tgt = a_ptr.add(block_idx(op.ti, op.tj) * block_size);
            match op.kind {
                0 => potrf_block(tgt, b),
                1 => trsm_block(tgt, a_ptr.add(block_idx(op.s1i, op.s1j) * block_size), b),
                2 => syrk_block(tgt, a_ptr.add(block_idx(op.s1i, op.s1j) * block_size), b),
                3 => gemm_block(tgt,
                                a_ptr.add(block_idx(op.s1i, op.s1j) * block_size),
                                a_ptr.add(block_idx(op.s2i, op.s2j) * block_size), b),
                _ => panic!("unknown op kind"),
            }
        }
    }
    let secs = t0.elapsed().as_secs_f64();

    // Checksum
    let mut cs: u64 = 0;
    for v in &a {
        let fixed = (*v * 1e6) as i64;
        cs = (cs + (fixed as u32 as u64)) & 0xFFFFFFFF_u64;
    }
    println!("CHECKSUM={}", cs);
    println!("RUNTIME_SEC={}", secs);
}
